### PR TITLE
Add .codegen.rewriting module

### PR DESCRIPTION
**Pull request by bjodah**
This module is useful for rewriting symbolic expressions to target specific floating point functions in e.g. C99.

### CODE REVIEW

File: sympy/codegen/rewriting.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+Classes and functions useful for rewriting expressions for optimized code
+generation. Some languages (or standards thereof), e.g. C99, offer specialized
+math functions for better performance and/or precision.
+
+Using the ``optimize`` function in this module, together with a collection of
+rules (represented as instances of ``Optimization``), one can rewrite the
+expressions for this purpose.
+
+Examples
+--------
```
Comment (smichr) [inline sympy/codegen/rewriting.py]: Should this be `========` with a blank line following?

File: sympy/codegen/rewriting.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+
+class Optimization(object):
+    """ Abstract base class for rewriting optimization.
+
+    Subclasses should implement ``__call__`` taking an expression
+    as argument.
+
+    Parameters
+    ----------
```
Comment (smichr) [inline sympy/codegen/rewriting.py]: `==========`?

Comment (smichr) [reply sympy/codegen/rewriting.py]: ditto below

**Comment by bjodah** [comment]
Thanks @smichr 
I've always been a bit confused regarding `====` vs. `-----` (numpydoc [guidelines](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) are in favor of the latter but you're right that SymPy usually uses the former).

I haven added the module to `doc/src/modules/codegen.rst` since that would just give a merge conflict with #13100, but once one of these are merged I'll merge master in and add it.

**Comment by asmeurer** [comment]
This looks like a good start.

**Pull request merged**